### PR TITLE
[Merged by Bors] - feat: results about `List.orderedInsert`

### PR DESCRIPTION
**Pull request by eric-wieser**
I suspect I will need some of these for #10660

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Sort.lean [hunk 1]
```diff
@@ -276,6 +276,44 @@ theorem Sorted.insertionSort_eq : ∀ {l : List α}, Sorted r l → insertionSor
...
+theorem erase_orderedInsert [DecidableEq α] [IsRefl α r] (x : α) (xs : List α) :
+    (xs.orderedInsert r x).erase x = xs := by
+  rw [orderedInsert_eq_take_drop, erase_append_right, List.erase_cons_head,
+    takeWhile_append_dropWhile]
+  intro h
+  replace h := mem_takeWhile_imp h
+  simp [refl x] at h
+
+/-- For an irreflexive relation, inserting then erasing is the identity. -/
+theorem erase_orderedInsert_of_nmem [DecidableEq α] [IsIrrefl α r]
```
Comment (eric-wieser) [inline Mathlib/Data/List/Sort.lean:290]: I can't remember if `nmem` is an allowed name here.

File: Mathlib/Data/List/Sort.lean [hunk 2]
```diff
@@ -276,6 +287,48 @@ theorem Sorted.insertionSort_eq : ∀ {l : List α}, Sorted r l → insertionSor
...
+theorem erase_orderedInsert [DecidableEq α] [IsRefl α r] (x : α) (xs : List α) :
+    (xs.orderedInsert r x).erase x = xs := by
+  rw [orderedInsert_eq_take_drop, erase_append_right, List.erase_cons_head,
+    takeWhile_append_dropWhile]
+  intro h
+  replace h := mem_takeWhile_imp h
+  simp [refl x] at h
+
+/-- Inserting then erasing an element that is absent is the identity. -/
+theorem erase_orderedInsert_of_nmem [DecidableEq α]
```
Comment (YaelDillies) [inline Mathlib/Data/List/Sort.lean:300]: Do we use `nmem` as a naming convention anywhere?

File: Mathlib/Data/List/Sort.lean [hunk 3]
```diff
@@ -276,6 +287,48 @@ theorem Sorted.insertionSort_eq : ∀ {l : List α}, Sorted r l → insertionSor
...
+    (x : α) (xs : List α) (hx : x ∉ xs) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Sort.lean:301]:
```suggestion
    {x : α} {xs : List α} (hx : x ∉ xs) :
```
is better probably?

File: Mathlib/Data/List/Sort.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Data/List/Sort.lean]: yes, but 10x less than `not_mem`, and not in the `List` namespace; edited.

**Comment by YaelDillies** [comment]
Thanks!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**